File: KAWebViewController/KAWebViewController/KAWebViewController.m
//
//  KAWViewController.m
//  KAWebViewController
//
//  Created by Kyle Adams on 08-04-14.
//  Copyright (c) 2014 Kyle Adams. All rights reserved.
//

#import "KAWebViewController.h"

#define FORWARD_BUTTON @"KAWForward"
#define BACK_BUTTON @"KAWBack"
#define IPAD UI_USER_INTERFACE_IDIOM() == UIUserInterfaceIdiomPad

@interface KAWebViewController () <UIWebViewDelegate>

@property (strong, nonatomic) UIWebView *webView;

@property (strong, nonatomic) UIBarButtonItem *space;
@property (strong, nonatomic) UIBarButtonItem *refreshButton;
@property (strong, nonatomic) UIBarButtonItem *backButton;
@property (strong, nonatomic) UIBarButtonItem *forwardButton;
@property (strong, nonatomic) UIBarButtonItem *stopButton;
@property (strong, nonatomic) UIBarButtonItem *actionButton;

@property (strong, nonatomic) NSArray *toolBarItemsWhenLoading;
@property (strong, nonatomic) NSArray *toolBarItemsWhenDoneLoading;

@end

@implementation KAWebViewController

- (instancetype)initWithURL:(NSURL *)url
{
    self = [super init];
    if (self) {
        self.url = url;
    }
    return self;
}

#pragma mark - Properties

- (NSArray *)toolBarItemsWhenLoading
{
    if (!_toolBarItemsWhenLoading) {
        if (IPAD) {
            _toolBarItemsWhenLoading = @[self.backButton,
                                         self.forwardButton,
                                         self.stopButton,
                                         self.actionButton];
        } else {
            _toolBarItemsWhenLoading = @[self.backButton,
                                         self.space,
                                         self.forwardButton,
                                         self.space,
                                         self.stopButton,
                                         self.space,
                                         self.actionButton];
        }
    }
    return _toolBarItemsWhenLoading;
}

- (NSArray *)toolBarItemsWhenDoneLoading
{
    if (!_toolBarItemsWhenDoneLoading) {
        if (IPAD) {
            _toolBarItemsWhenDoneLoading = @[self.backButton,
                                         self.forwardButton,
                                         self.refreshButton,
                                         self.actionButton];
        } else {
            _toolBarItemsWhenDoneLoading = @[self.backButton,
                                         self.space,
                                         self.forwardButton,
                                         self.space,
                                         self.refreshButton,
                                         self.space,
                                         self.actionButton];
        }
    }
    return _toolBarItemsWhenDoneLoading;
}

- (UIBarButtonItem *)space
{
    if (!_space) {
        _space = [[UIBarButtonItem alloc] initWithBarButtonSystemItem:UIBarButtonSystemItemFlexibleSpace target:self action:nil];
    }
    return _space;
}

- (UIBarButtonItem *)backButton
{
    if (!_backButton) {
        _backButton = [[UIBarButtonItem alloc] initWithImage:[UIImage imageNamed:BACK_BUTTON]
                                                       style:UIBarButtonItemStylePlain
                                                      target:self
                                                      action:@selector(previousPage)];
    }
    return _backButton;
}

- (UIBarButtonItem *)forwardButton
{
    if (!_forwardButton) {
        _forwardButton = [[UIBarButtonItem alloc] initWithImage:[UIImage imageNamed:FORWARD_BUTTON]
                                                          style:UIBarButtonItemStylePlain
                                                         target:self
                                                         action:@selector(forwardPage)];
    }
    return _forwardButton;
}

- (UIBarButtonItem *)refreshButton
{
    if (!_refreshButton) {
        _refreshButton = [[UIBarButtonItem alloc] initWithBarButtonSystemItem:UIBarButtonSystemItemRefresh
                                                                       target:self
                                                                       action:@selector(refreshPage)];
    }
    return _refreshButton;
}

- (UIBarButtonItem *)stopButton
{
    if (!_stopButton) {
        _stopButton = [[UIBarButtonItem alloc] initWithBarButtonSystemItem:UIBarButtonSystemItemStop
                                                                    target:self
                                                                    action:@selector(stopRefresh)];
    }
    return _stopButton;
}

- (UIBarButtonItem *)actionButton
{
    if (!_actionButton) {
        _actionButton = [[UIBarButtonItem alloc] initWithBarButtonSystemItem:UIBarButtonSystemItemAction
                                                                      target:self
                                                                      action:@selector(actionPressed)];
    }
    return _actionButton;
}

- (void)setUrl:(NSURL *)url
{
    _url = url;
    self.webView.delegate = self;
    self.webView.scalesPageToFit = YES;
    [self.webView loadRequest:[NSURLRequest requestWithURL:self.url]];
}

- (UIWebView *)webView
{
    if (!_webView) {
        _webView = [[UIWebView alloc] init];
    }
    return _webView;
}

#pragma mark - Target Actions

-(void)previousPage
{
    [self.webView goBack];
}

-(void)forwardPage
{
    [self.webView goForward];
}

- (void)refreshPage
{
    [self.webView reload];
}

- (void)stopRefresh
{
    [self.webView stopLoading];
    [self updateUI];
}

- (void)actionPressed
{
    NSArray *actionItems = @[self.webView.request.URL];
    UIActivityViewController *avc = [[UIActivityViewController alloc] initWithActivityItems:actionItems applicationActivities:nil];
    
    [self presentViewController:avc animated:YES completion:nil];
    

}

#pragma mark - UIViewController Lifecycle

- (void)viewDidLoad
{
    [super viewDidLoad];
    self.navigationController.toolbar.tintColor = self.navigationController.navigationBar.tintColor;
}

- (void)viewDidAppear:(BOOL)animated
{
    [super viewDidAppear:animated];
    //for a snappy UI
    self.view = self.webView;
}

- (void)updateUI
{
    [self setToolbarItemsForState:self.webView.loading];
    
    self.backButton.enabled = self.webView.canGoBack ? YES : NO;
    self.forwardButton.enabled = self.webView.canGoForward? YES : NO;
    
    if (!IPAD) {
        if (self.navigationController.toolbar.hidden) {
            [self.navigationController setToolbarHidden:NO animated:NO];
        }
    }
}

- (void)setToolbarItemsForState:(BOOL)loading
{
    if (loading) {
        if (!IPAD) {
            self.toolbarItems = self.toolBarItemsWhenLoading;
        } else {
            self.navigationItem.rightBarButtonItems = self.toolBarItemsWhenLoading.reverseObjectEnumerator.allObjects;
        }
        
    } else {
        
        if (!IPAD) {
            self.toolbarItems = self.toolBarItemsWhenDoneLoading;
        } else {
            self.navigationItem.rightBarButtonItems = self.toolBarItemsWhenDoneLoading.reverseObjectEnumerator.allObjects;
        }
        
        self.title = [self.webView stringByEvaluatingJavaScriptFromString:@"document.title"];

    }
}

- (void)viewWillDisappear:(BOOL)animated
{
    [self.navigationController setToolbarHidden:YES animated:YES];
}

#pragma mark - UIWebViewDelegate

- (void)webViewDidStartLoad:(UIWebView *)webView
{
    [[UIApplication sharedApplication] setNetworkActivityIndicatorVisible:YES];
    [self updateUI];
}

- (void)webViewDidFinishLoad:(UIWebView *)webView
{
    [[UIApplication sharedApplication] setNetworkActivityIndicatorVisible:NO];
    [self updateUI];
}

- (void)webView:(UIWebView *)webView didFailLoadWithError:(NSError *)error
{
    [[UIApplication sharedApplication] setNetworkActivityIndicatorVisible:NO];
    [self updateUI];
}

@end

File: Example/KAWebViewController.swift
//
//  KAWebViewController.swift
//  KAWebViewController
//
//  Very basic swift version of KAWWebView, currently only works with Xcode 6 beta.
//  Created by Kyle Adams on 04-06-14.
//  Copyright (c) 2014 Kyle Adams. All rights reserved.
//

import UIKit

class KAWebViewController: UIViewController, UIWebViewDelegate {
    
    let webView = UIWebView()
    let refreshButton = UIBarButtonItem(barButtonSystemItem: .Refresh,target: nil, action: "reloadPage")
    let backButton = UIBarButtonItem(image: UIImage(named: "KAWBack"), style: .Plain, target: nil, action: "goBack")
    let forwardButton = UIBarButtonItem(image: UIImage(named: "KAWForward"), style: .Plain, target: nil, action: "goForward")
    let actionButton = UIBarButtonItem(barButtonSystemItem: .Action, target: nil, action: "showActionMenu")
    let stopButton = UIBarButtonItem(barButtonSystemItem: .Stop, target: nil, action: "stopLoading")
    let space = UIBarButtonItem(barButtonSystemItem: .FlexibleSpace, target: nil, action: nil)
    
    var url: NSURL? {
    didSet {
        loadWebView()
    }
    }
    

    override func viewDidLoad() {
        super.viewDidLoad()

        // Do any additional setup after loading the view.
        loadToolBarItems()
        navigationController.toolbarHidden = false
        webView.delegate = self
        webView.scalesPageToFit = true
        view = webView
        
        updateUI()
    }
    
    override func viewWillDisappear(animated: Bool) {
        navigationController.toolbarHidden = navigationController.toolbarHidden ? false : true
    }
    
    func updateUI() {
        setToolBarItemsForState(webView.loading)
        
        backButton.enabled = webView.canGoBack ? true : false
        forwardButton.enabled = webView.canGoForward ? true : false
        
        if navigationController.toolbarHidden {
            navigationController.toolbarHidden = false
        }
    }
    
    func setToolBarItemsForState(loading: Bool) {
        
        if loading {
            //set toolbar items when loading
            println("WebView loading!")
            toolbarItems = [backButton, space, forwardButton, space, stopButton, space, actionButton]
        } else {
            //set toolbar items when not loading
            println("WebView is not loading")
            toolbarItems = [backButton, space, forwardButton, space, refreshButton, space, actionButton]
        }
    }
    
    func loadToolBarItems() {
        refreshButton.target = self
        backButton.target = self
        forwardButton.target = self
        actionButton.target = self
        stopButton.target = self
    }
    
    func loadWebView() {
        let request = NSURLRequest(URL: url!)
        webView.loadRequest(request)
    }
    
    func goBack() {
        if webView.canGoBack {
            webView.goBack()
        }
    }
    
    func goForward() {
        if webView.canGoForward {
            webView.goForward()
        }
    }
    
    func reloadPage() {
        webView.reload()
    }
    
    func stopLoading() {
        webView.stopLoading()
        updateUI()
    }
    
    func showActionMenu() {
        let avc = UIActivityViewController(activityItems: [webView.request.URL], applicationActivities: nil)
        self.presentViewController(avc, animated: false, completion: nil)
    }
    
    func webViewDidStartLoad(_: UIWebView!) {
        UIApplication.sharedApplication().networkActivityIndicatorVisible = true
        updateUI()
    }
    
    func webViewDidFinishLoad(_:UIWebView!) {
        UIApplication.sharedApplication().networkActivityIndicatorVisible = false
        updateUI()
    }
    
    func webView(_: UIWebView!, didFailLoadWithError error: NSError!) {
        UIApplication.sharedApplication().networkActivityIndicatorVisible = false
        updateUI()
    }
}
